src/lib/support/value_parser/ValueParserComplex.h

```diff
@@ -0,0 +1,147 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <charconv>
+#include <functional>
+#include <type_traits>
+#include <vector>
+
+#include <app/data-model/List.h>
+#include <app/data-model/Nullable.h>
+#include <lib/core/CHIPError.h>
+#include <lib/support/value_parser/ValueParserIntegral.h>
+
+namespace chip {
+
+template <typename T>
+CHIP_ERROR ParseValue(const char *& iter, const char * end, chip::Optional<T> & value,
+                      std::vector<std::function<void(void)>> & freeFunctions);
+template <typename T>
+CHIP_ERROR ParseValue(const char *& iter, const char * end, chip::app::DataModel::Nullable<T> & value,
+                      std::vector<std::function<void(void)>> & freeFunctions);
+template <typename ListEntryType>
+CHIP_ERROR ParseValue(const char *& iter, const char * end, chip::app::DataModel::List<ListEntryType> & value,
+                      std::vector<std::function<void(void)>> & freeFunctions);
+
+template <typename T>
+CHIP_ERROR ParseValue(const char *& iter, const char * end, chip::Optional<T> & value,
+                      std::vector<std::function<void(void)>> & freeFunctions)
+{
+    iter = SkipSpaces(iter);
```

Comment:
This is pretty confusing: this seems to be assuming `iter` is null-terminated, in which case why are we carrying around `end`?  And if `end` is really needed, then `SkipSpaces`, and any operation that increments `iter`, should get `end` handed to it.